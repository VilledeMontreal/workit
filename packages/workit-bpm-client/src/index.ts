/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

// IOC
export * from './config/constants/identifiers';
import './config/container';

export * from './camundaBpmClient';
export * from './bpmClient';
export * from './bpmLogger';
export * from './camundaBpmClient';
export * from './camundaMapperProperties';
export * from './logger';
export * from './camundaMessage';

export * from './utils/utils';
export * from './variables';
