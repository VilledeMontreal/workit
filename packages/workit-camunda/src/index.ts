// Copyright (c) Ville de Montreal. All rights reserved.
// Licensed under the MIT license.
// See LICENSE file in the project root for full license information.

import './config/ioc';

export * from './config/constants';

// IOC
export * from './config/constants/identifiers';
export * from './config/constants/tag';
export * from './config/ioc';
export * from './models/IoC';

// Camunda specific
// // specs
export * from './models/camunda/specs/camundaClient';
export * from './models/camunda/specs/camundaConfig';
export * from './models/camunda/specs/subscriptionOptions';
export * from './models/camunda/specs/payload';
// // General
export * from './models/camunda/utils';
export * from './models/camunda/variables';
export * from './models/camunda/camundaBpmClient';
export * from './models/camunda/camundaMapperProperties';
export * from './models/camunda/camundaMessage';
export * from './models/camunda/logger';

// Zeebe specific
// // specs
export * from './models/zeebe/specs/payload';
// export * from './models/zeebe/specs/workflowDeployResponse';
export * from './models/zeebe/specs/zeebeClientOptions';
export * from './models/zeebe/specs/zeebeOptions';
export * from './models/zeebe/specs/zeebeWorkerOptions';
export * from './models/zeebe/specs/logLevel';
// // General
export * from './models/zeebe/zeebeClient';
export * from './models/zeebe/zeebeMapperProperties';
export * from './models/zeebe/zeebeMessage';

// Client lib
export * from './models/camunda-n-mq/specs/incidentException';
export * from './models/camunda-n-mq/specs/failureException';
export * from './models/camunda-n-mq/specs/camundaService';
export * from './models/camunda-n-mq/specs/client';
export * from './models/camunda-n-mq/specs/workflowClient';
export * from './models/camunda-n-mq/specs/message';
export * from './models/camunda-n-mq/specs/properties';
export * from './models/camunda-n-mq/specs/propertiesBase';
export * from './models/camunda-n-mq/specs/customHeaders';
export * from './models/camunda-n-mq/specs/createWorkflowInstance';
export * from './models/camunda-n-mq/specs/createWorkflowInstanceResponse';
export * from './models/camunda-n-mq/specs/deployWorkflowResponse';
export * from './models/camunda-n-mq/specs/publishMessage';
export * from './models/camunda-n-mq/specs/updateWorkflowVariables';
export * from './models/camunda-n-mq/specs/workflow';
export * from './models/camunda-n-mq/specs/workflowResponse';
export * from './models/camunda-n-mq/client';
export * from './models/camunda-n-mq/clientManager';
export * from './utils/utils';

// Core
export * from './models/core/processHandler/specs/processHandlerConfig';
export * from './models/core/specs/process';
export * from './models/core/specs/task';
export * from './models/core/specs/failureStrategy';
export * from './models/core/specs/successStrategy';
export * from './models/core/worker';
export * from './models/core/specs/taskBase';
export * from './models/core/specs/interceptor';

// Core instrumentation
export * from './models/core/instrumentations/specs/camundaClientInstrumentation';
export * from './models/core/instrumentations/specs/camundaClientTracer';
export * from './models/core/instrumentations/specs/instrumentation';
export * from './models/core/instrumentations/camundaClientTracer';
export * from './models/core/instrumentations/enums/apm';
