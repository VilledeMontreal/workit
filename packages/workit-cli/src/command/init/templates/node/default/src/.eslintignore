node_modules
dist
.vscode
coverage
*lock*
*.md
tests