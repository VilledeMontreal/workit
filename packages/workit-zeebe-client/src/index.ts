/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

// IOC
export * from './config/constants/identifiers';
import './config/container';

export * from './zeebeClient';
export * from './zeebeMapperProperties';
export * from './zeebeMessage';
