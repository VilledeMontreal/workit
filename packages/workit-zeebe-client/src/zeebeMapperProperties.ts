/*
 * Copyright (c) 2020 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */
import { IEmptyPayload, IPayload, IWorkflowProps } from 'workit-types';

export class ZeebeMapperProperties {
  public static map<TVariables = unknown, TProps = unknown>(obj: IPayload<TVariables, TProps>): IWorkflowProps<TProps> {
    let businessKey;
    if (obj.variables || (obj.variables as any).businessKey) {
      businessKey = (obj.variables as any).businessKey;
    }

    return {
      businessKey, // retro compatibility with bpmn workflow engine
      activityId: obj.elementId,
      processInstanceId: obj.elementInstanceKey,
      workflowDefinitionVersion: obj.workflowDefinitionVersion,
      workflowInstanceKey: obj.workflowInstanceKey,
      workflowKey: obj.workflowKey,
      bpmnProcessId: obj.bpmnProcessId,
      customHeaders: { ...obj.customHeaders },
      jobKey: obj.key,
      retries: obj.retries,
      topicName: obj.type,
      workerId: obj.worker,
      lockExpirationTime: new Date(Number(obj.deadline))
    };
  }

  public static unmap<TProps = unknown>(props: IWorkflowProps<TProps>): IEmptyPayload<TProps> {
    return {
      elementId: props.activityId,
      elementInstanceKey: props.processInstanceId,
      workflowKey: props.workflowKey,
      workflowInstanceKey: props.workflowInstanceKey,
      workflowDefinitionVersion: props.workflowDefinitionVersion,
      bpmnProcessId: props.bpmnProcessId,
      retries: props.retries,
      worker: props.workerId,
      deadline: props.lockExpirationTime.getTime().toString(),
      customHeaders: { ...props.customHeaders },
      key: props.jobKey,
      type: props.topicName
    };
  }
}
