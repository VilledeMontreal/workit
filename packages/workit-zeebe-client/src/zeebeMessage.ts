/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { getVariablesWhenChanged } from 'workit-core';
import { FailureException, ICamundaService, IMessage, IPayload, IWorkflowProps } from 'workit-types';
// FIXME:dist folder....
import { CompleteFn } from 'zeebe-node/dist/lib/interfaces';
import { ZeebeMapperProperties } from './zeebeMapperProperties';

export class ZeebeMessage {
  public static wrap<TVariables, TProps>(
    payload: IPayload<TVariables, TProps>,
    complete: CompleteFn<TVariables>
  ): [IMessage<TVariables, IWorkflowProps<TProps>>, ICamundaService] {
    const properties = ZeebeMapperProperties.map(payload);
    return [
      {
        body: payload.variables,
        properties
      },
      {
        hasBeenThreated: false,
        async ack(message: IMessage<Partial<TVariables> | undefined, IWorkflowProps<unknown>>) {
          if (this.hasBeenThreated) {
            return Promise.resolve();
          }

          // TODO: change any to real type body
          const vars = getVariablesWhenChanged<any>(message, msg => ZeebeMessage.unwrap(msg));

          this.hasBeenThreated = await complete.success(vars.variables);
        },
        nack(error: FailureException) {
          if (this.hasBeenThreated) {
            return Promise.resolve();
          }
          const retries = error.retries;
          // TODO: check if zeebe-node made the type correction
          this.hasBeenThreated = (complete.failure(error.message, retries) as unknown) as boolean;
          return Promise.resolve();
        }
      }
    ];
  }

  /**
   * Shallow copy
   */
  public static unwrap<TVariables, TProps>(
    message: IMessage<TVariables, IWorkflowProps<TProps>>
  ): IPayload<TVariables, TProps> {
    const emptyPayload = ZeebeMapperProperties.unmap<TProps>(message.properties);
    (emptyPayload as IPayload<TVariables, TProps>).variables = message.body;
    return emptyPayload as IPayload<TVariables, TProps>;
  }
}
