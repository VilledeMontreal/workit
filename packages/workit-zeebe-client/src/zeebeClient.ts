/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */
import { optional } from 'inversify';
import {
  ICamundaService,
  IClient,
  ICreateWorkflowInstance,
  ICreateWorkflowInstanceResponse,
  IDeployWorkflowResponse,
  IMessage,
  IPagination,
  IPaginationOptions,
  IPayload,
  IPublishMessage,
  IUpdateWorkflowRetry,
  IUpdateWorkflowVariables,
  IWorkflow,
  IWorkflowClient,
  IWorkflowDefinition,
  IWorkflowDefinitionKey,
  IWorkflowDefinitionRequest,
  IWorkflowOptions,
  IWorkflowProcessIdDefinition,
  IWorkflowProps,
  IZeebeOptions
} from 'workit-types';
import { Configs, IAPIConfig as IElasticExporterConfig, ZBElasticClient } from 'zeebe-elasticsearch-client';
import { ZBClient, ZBWorker } from 'zeebe-node';
// FIXME: dist folder
import { CompleteFn } from 'zeebe-node/dist/lib/interfaces';
import { PaginationUtils } from './utils/paginationUtils';
import { ZeebeMessage } from './zeebeMessage';

export class ZeebeClient<TVariables = unknown, TProps = unknown, RVariables = TVariables>
  implements IClient<ICamundaService>, IWorkflowClient {
  private readonly _client: ZBClient;
  private readonly _exporterClient: ZBElasticClient;
  private _worker: ZBWorker<TVariables, TProps, RVariables> | undefined;
  private readonly _config: IZeebeOptions;
  private readonly _exporterConfig: Partial<IElasticExporterConfig> | undefined;
  constructor(config: IZeebeOptions, client?: ZBClient, exporterConfig?: Partial<IElasticExporterConfig>);
  constructor(config: IZeebeOptions, exporterConfig?: Partial<IElasticExporterConfig>);
  constructor(
    config: IZeebeOptions,
    @optional() client?: ZBClient | Partial<IElasticExporterConfig>,
    @optional() exporterConfig?: Partial<IElasticExporterConfig>
  ) {
    this._config = config;

    if (client && (client as Partial<IElasticExporterConfig>).url) {
      this._exporterConfig = client as Partial<IElasticExporterConfig>;
      this._client = new ZBClient(config.baseUrl, config);
    } else {
      this._client = (client || new ZBClient(config.baseUrl, config)) as ZBClient;
      this._exporterConfig = exporterConfig;
    }

    if (!exporterConfig) {
      // tslint:disable-next-line: no-console
      console.log(
        "warning: no exporterConfig has been provided to Zeebe. getWorkflow and getWorkflows methods won't work. "
      );
    }
    this._exporterClient = new ZBElasticClient(new Configs(exporterConfig));
  }
  public subscribe(
    onMessageReceived: (
      message: IMessage<TVariables, IWorkflowProps<TProps>>,
      service: ICamundaService
    ) => Promise<void>
  ): Promise<void> {
    this._worker = this._client.createWorker(
      this._config.workerId || 'some-random-id',
      this._config.topicName,
      async (payload: IPayload<TVariables, TProps>, complete: CompleteFn<TVariables>) => {
        const [message, service] = ZeebeMessage.wrap<TVariables, TProps>(payload, complete);
        await onMessageReceived(message, service);
      },
      this._config
    );
    return Promise.resolve();
  }
  public async deployWorkflow(bpmnPath: string): Promise<IDeployWorkflowResponse> {
    const result = await this._client.deployWorkflow(bpmnPath);
    return {
      workflows: result.workflows,
      key: result.key.toString() // TODO: interface say number but it return string, need to PR to zeebe-node
    };
  }
  public async getWorkflows(options?: Partial<IWorkflowOptions & IPaginationOptions>): Promise<IPagination<IWorkflow>> {
    this._validateExporterConfig();
    const params = { _source_excludes: 'bpmnXml' };
    const workflowOptions = { params };
    workflowOptions.params = PaginationUtils.setElasticPaginationParams(params, options);
    const criteria = this._setWorkflowCriteria(options);
    const result = await this._exporterClient.getWorkflows(criteria, workflowOptions);
    const elasticResult = result.data.hits;
    const data = elasticResult.hits.map(doc => {
      const workflow = doc._source;
      return {
        bpmnProcessId: workflow.bpmnProcessId,
        version: workflow.version,
        workflowKey: workflow.key.toString(),
        resourceName: workflow.resourceName
      };
    });

    return {
      paging: PaginationUtils.getPagingFromOptions(elasticResult.total, options),
      items: data
    };
  }
  public async getWorkflow(payload: IWorkflowDefinitionRequest): Promise<IWorkflowDefinition> {
    this._validateExporterConfig();
    this._validateObject();
    const key = Number((payload as IWorkflowDefinitionKey).workflowKey);
    const response = await this._exporterClient.getWorkflows({
      key: !isNaN(key) ? key : undefined,
      bpmnProcessId: (payload as IWorkflowProcessIdDefinition).bpmnProcessId,
      version: (payload as IWorkflowProcessIdDefinition).version,
      latestVersion: typeof (payload as IWorkflowProcessIdDefinition).version !== 'number'
    });

    const data = response.data.hits.hits[0];

    if (!data) {
      throw new Error('Not found');
    }

    const doc = data._source;

    return {
      version: doc.version,
      resourceName: doc.resourceName,
      bpmnXml: doc.bpmnXml,
      workflowKey: doc.key.toString(),
      bpmnProcessId: doc.bpmnProcessId
    };
  }
  public updateVariables<T = unknown>(model: IUpdateWorkflowVariables<Partial<T>>): Promise<void> {
    return this._client.setVariables<T>({
      elementInstanceKey: model.processInstanceId,
      variables: model.variables,
      local: !!model.local
    });
  }
  public updateJobRetries(payload: IUpdateWorkflowRetry): Promise<void> {
    return this._client.updateJobRetries(payload);
  }
  /**
   * Publish a message event
   * If you don't specify correlationKey, message will be treated as message start event
   */
  public publishMessage<T = unknown>(payload: IPublishMessage<T, string>): Promise<void> {
    return this._client.publishMessage({
      correlationKey: payload.correlation || '__MESSAGE_START_EVENT__',
      variables: payload.variables,
      messageId: payload.messageId,
      timeToLive: payload.timeToLive,
      name: payload.name
    });
  }
  public createWorkflowInstance<T = unknown>(
    model: ICreateWorkflowInstance<T>
  ): Promise<ICreateWorkflowInstanceResponse> {
    if (!model.version) {
      return this._client.createWorkflowInstance<T>(model.bpmnProcessId, model.variables);
    }
    return this._client.createWorkflowInstance<T>(model as Required<ICreateWorkflowInstance<T>>);
  }
  public cancelWorkflowInstance(instance: string): Promise<void> {
    this._validateNumber(instance);
    return this._client.cancelWorkflowInstance(instance as any); // TODO: will be fixed https://github.com/zeebe-io/zeebe/issues/2680
  }
  public resolveIncident(incidentKey: string): Promise<void> {
    return this._client.resolveIncident(incidentKey);
  }
  public unsubscribe(): Promise<void> {
    if (this._worker) {
      return this._worker.close();
    }
    return Promise.resolve();
  }

  private _validateExporterConfig() {
    if (!this._exporterConfig) {
      throw new Error(`
      Please, refer to the warning when you instiate this class. You must pass exporterConfig to the Ctor in order to use this method.
      For now, we are only compatible with elastic indexes that operate produce. If you use different indexes or different exporter, please raise an issue.
      `);
    }
  }
  private _validateObject() {
    if (!this._exporterConfig) {
      throw new Error(`
        Object passed to the method can't be undefined
      `);
    }
  }
  private _validateNumber(variable: string) {
    const value = Number(variable);
    if (!Number.isInteger(value)) {
      throw new Error(`
      workflowInstanceKey value is malformed
      `);
    }
  }

  private _setWorkflowCriteria(options?: Partial<IWorkflowOptions & IPaginationOptions>) {
    if (!options) {
      return {};
    }
    return { bpmnProcessId: options.bpmnProcessId };
  }
}
