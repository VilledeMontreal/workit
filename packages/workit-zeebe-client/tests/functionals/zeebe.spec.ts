/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import * as nock from 'nock';
import { FailureStrategySimple, SCProcessHandler, SuccessStrategySimple, Worker } from 'workit-core';
import { ICamundaConfig } from 'workit-types';
import { ZeebeClient } from '../../src/zeebeClient';

process.env.ZB_NODE_LOG_LEVEL = process.env.ZB_NODE_LOG_LEVEL || 'NONE';
const run = (worker: Worker, done: any, delay = 500) => {
  worker.start();
  worker.run();

  setTimeout(async () => {
    await worker.stop();
    done();
  }, delay);
};

describe('Zeebe Worker', () => {
  let worker: Worker;
  let config: ICamundaConfig;

  beforeEach(() => {
    config = {
      maxTasks: 1,
      workerId: 'test-worker',
      baseUrl: 'localhost:26500',
      topicName: 'demo-service',
      bpmnKey: 'test'
    };
  });

  it('should instanciate the worker', done => {
    (async () => {
      const zeebeClient = new ZeebeClient(config);
      const successHandler = new SuccessStrategySimple();
      const failureHandler = new FailureStrategySimple();
      successHandler.handle = jest.fn().mockResolvedValueOnce({});
      const processHandler = new SCProcessHandler(successHandler, failureHandler, {} as any);
      worker = new Worker(zeebeClient, processHandler);
      run(worker, done, 500);
    })().catch();
  });

  it('should get workflows', async () => {
    const scope = nock('http://localhost:9200')
      .post('/operate-workflow_alias/_search')
      .query({ _source_excludes: 'bpmnXml' })
      .reply(200, require(`${process.cwd()}/tests/data/elasticResponse.workflow`));
    const zeebeClient = new ZeebeClient(config, undefined, {
      url: 'http://localhost:9200'
    });
    const response = await zeebeClient.getWorkflows();

    scope.done();
    expect(response).toMatchSnapshot();
  });

  it('should get MESSAGE_EVENT workflow', async () => {
    const scope = nock('http://localhost:9200')
      .post('/operate-workflow_alias/_search')
      .reply(200, require(`${process.cwd()}/tests/data/elasticResponseAgg.workflow`));

    const zeebeClient = new ZeebeClient(config, undefined, {
      url: 'http://localhost:9200'
    });
    const response = await zeebeClient.getWorkflow({ bpmnProcessId: 'MESSAGE_EVENT' });

    scope.done();
    expect(response).toMatchSnapshot();
  });

  it('should generate an exception for cancelWorkflowInstance when key is malformed', async () => {
    const zeebeClient = new ZeebeClient(config);
    try {
      await zeebeClient.cancelWorkflowInstance('hello');
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
  it('should limit the number of document', async () => {
    const size = 5;
    const scope = nock('http://localhost:9200')
      .post('/operate-workflow_alias/_search', { query: { bool: { must: [] } } })
      .query({ _source_excludes: 'bpmnXml', size })
      .reply(200, require(`${process.cwd()}/tests/data/elasticResponse.paginated`));

    const zeebeClient = new ZeebeClient(config, {
      url: 'http://localhost:9200'
    });
    const response = await zeebeClient.getWorkflows({ size });
    scope.done();
    expect(response.items.length).toEqual(size);
    expect(response).toMatchSnapshot();
  });

  it('should limit the number of document and skip one document', async () => {
    const size = 4;
    const from = 1;
    const scope = nock('http://localhost:9200')
      .post('/operate-workflow_alias/_search', { query: { bool: { must: [] } } })
      .query({ _source_excludes: 'bpmnXml', size, from })
      .reply(200, require(`${process.cwd()}/tests/data/elasticResponse.paginated.skip`));

    const zeebeClient = new ZeebeClient(config, {
      url: 'http://localhost:9200'
    });
    const response = await zeebeClient.getWorkflows({ size, from });
    scope.done();
    expect(response.items.length).toEqual(size);
    expect(response).toMatchSnapshot();
  });

  it('should limit the number of document and search by bpmnProcessId', async () => {
    const size = 5;
    const bpmnProcessId = 'MESSAGE_EVENT';
    const scope = nock('http://localhost:9200')
      .post('/operate-workflow_alias/_search', {
        query: { bool: { must: [{ match: { bpmnProcessId: { query: bpmnProcessId } } }] } }
      })
      .query({ _source_excludes: 'bpmnXml', size })
      .reply(200, require(`${process.cwd()}/tests/data/elasticResponseBpmnProcessId.paginated`));

    const zeebeClient = new ZeebeClient(config, {
      url: 'http://localhost:9200'
    });
    const response = await zeebeClient.getWorkflows({ size, bpmnProcessId });
    scope.done();
    expect(response).toMatchSnapshot();
  });
});
