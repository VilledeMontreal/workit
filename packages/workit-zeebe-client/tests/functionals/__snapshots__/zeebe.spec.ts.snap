// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zeebe Worker should generate an exception for cancelWorkflowInstance when key is malformed 1`] = `
[Error: 
      workflowInstanceKey value is malformed
      ]
`;

exports[`Zeebe Worker should get MESSAGE_EVENT workflow 1`] = `
Object {
  "bpmnProcessId": "MESSAGE_EVENT",
  "bpmnXml": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<bpmn:definitions xmlns:bpmn=\\"http://www.omg.org/spec/BPMN/20100524/MODEL\\" xmlns:bpmndi=\\"http://www.omg.org/spec/BPMN/20100524/DI\\" xmlns:dc=\\"http://www.omg.org/spec/DD/20100524/DC\\" xmlns:zeebe=\\"http://camunda.org/schema/zeebe/1.0\\" xmlns:di=\\"http://www.omg.org/spec/DD/20100524/DI\\" id=\\"Definitions_00osekx\\" targetNamespace=\\"http://bpmn.io/schema/bpmn\\" exporter=\\"Zeebe Modeler\\" exporterVersion=\\"0.6.2\\">
  <bpmn:process id=\\"MESSAGE_EVENT\\" name=\\"MESSAGE EVENT\\" isExecutable=\\"true\\">
    <bpmn:startEvent id=\\"StartEvent_1\\" name=\\"start event\\">
      <bpmn:outgoing>SequenceFlow_1k0o5cy</bpmn:outgoing>
      <bpmn:messageEventDefinition messageRef=\\"Message_1l6q1nh\\" />
    </bpmn:startEvent>
    <bpmn:serviceTask id=\\"hello\\" name=\\"Hello\\">
      <bpmn:extensionElements>
        <zeebe:taskDefinition type=\\"topic_demo\\" retries=\\"5\\" />
      </bpmn:extensionElements>
      <bpmn:incoming>SequenceFlow_1k0o5cy</bpmn:incoming>
      <bpmn:outgoing>SequenceFlow_094m070</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:sequenceFlow id=\\"SequenceFlow_1k0o5cy\\" sourceRef=\\"StartEvent_1\\" targetRef=\\"hello\\" />
    <bpmn:boundaryEvent id=\\"BoundaryEvent_184rsr7\\" name=\\"catching\\" attachedToRef=\\"hello\\">
      <bpmn:outgoing>SequenceFlow_0umkghz</bpmn:outgoing>
      <bpmn:messageEventDefinition messageRef=\\"Message_1oxctc1\\" />
    </bpmn:boundaryEvent>
    <bpmn:serviceTask id=\\"ServiceTask_1vxgnug\\" name=\\"World\\">
      <bpmn:extensionElements>
        <zeebe:taskDefinition type=\\"topic_demo\\" retries=\\"1\\" />
      </bpmn:extensionElements>
      <bpmn:incoming>SequenceFlow_0umkghz</bpmn:incoming>
      <bpmn:outgoing>SequenceFlow_1w1cqrc</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:sequenceFlow id=\\"SequenceFlow_0umkghz\\" sourceRef=\\"BoundaryEvent_184rsr7\\" targetRef=\\"ServiceTask_1vxgnug\\" />
    <bpmn:endEvent id=\\"EndEvent_021uygl\\">
      <bpmn:incoming>SequenceFlow_094m070</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:sequenceFlow id=\\"SequenceFlow_094m070\\" sourceRef=\\"hello\\" targetRef=\\"EndEvent_021uygl\\" />
    <bpmn:endEvent id=\\"EndEvent_1o5r87f\\">
      <bpmn:incoming>SequenceFlow_1w1cqrc</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:sequenceFlow id=\\"SequenceFlow_1w1cqrc\\" sourceRef=\\"ServiceTask_1vxgnug\\" targetRef=\\"EndEvent_1o5r87f\\" />
  </bpmn:process>
  <bpmn:message id=\\"Message_1oxctc1\\" name=\\"catching\\">
    <bpmn:extensionElements>
      <zeebe:subscription correlationKey=\\"amount\\" />
    </bpmn:extensionElements>
  </bpmn:message>
  <bpmn:message id=\\"Message_1l6q1nh\\" name=\\"_MESSAGE_START_EVENT_\\" />
  <bpmndi:BPMNDiagram id=\\"BPMNDiagram_1\\">
    <bpmndi:BPMNPlane id=\\"BPMNPlane_1\\" bpmnElement=\\"MESSAGE_EVENT\\">
      <bpmndi:BPMNShape id=\\"StartEvent_0ibdhxr_di\\" bpmnElement=\\"StartEvent_1\\">
        <dc:Bounds x=\\"179\\" y=\\"103\\" width=\\"36\\" height=\\"36\\" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x=\\"171\\" y=\\"146\\" width=\\"52\\" height=\\"14\\" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id=\\"ServiceTask_14pv7xx_di\\" bpmnElement=\\"hello\\">
        <dc:Bounds x=\\"265\\" y=\\"81\\" width=\\"100\\" height=\\"80\\" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id=\\"SequenceFlow_1k0o5cy_di\\" bpmnElement=\\"SequenceFlow_1k0o5cy\\">
        <di:waypoint x=\\"215\\" y=\\"121\\" />
        <di:waypoint x=\\"265\\" y=\\"121\\" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id=\\"BoundaryEvent_184rsr7_di\\" bpmnElement=\\"BoundaryEvent_184rsr7\\">
        <dc:Bounds x=\\"347\\" y=\\"143\\" width=\\"36\\" height=\\"36\\" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x=\\"347\\" y=\\"183\\" width=\\"41\\" height=\\"14\\" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id=\\"ServiceTask_1vxgnug_di\\" bpmnElement=\\"ServiceTask_1vxgnug\\">
        <dc:Bounds x=\\"433\\" y=\\"201\\" width=\\"100\\" height=\\"80\\" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id=\\"SequenceFlow_0umkghz_di\\" bpmnElement=\\"SequenceFlow_0umkghz\\">
        <di:waypoint x=\\"365\\" y=\\"179\\" />
        <di:waypoint x=\\"365\\" y=\\"241\\" />
        <di:waypoint x=\\"433\\" y=\\"241\\" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id=\\"EndEvent_021uygl_di\\" bpmnElement=\\"EndEvent_021uygl\\">
        <dc:Bounds x=\\"415\\" y=\\"103\\" width=\\"36\\" height=\\"36\\" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id=\\"SequenceFlow_094m070_di\\" bpmnElement=\\"SequenceFlow_094m070\\">
        <di:waypoint x=\\"365\\" y=\\"121\\" />
        <di:waypoint x=\\"415\\" y=\\"121\\" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id=\\"EndEvent_1o5r87f_di\\" bpmnElement=\\"EndEvent_1o5r87f\\">
        <dc:Bounds x=\\"583\\" y=\\"223\\" width=\\"36\\" height=\\"36\\" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id=\\"SequenceFlow_1w1cqrc_di\\" bpmnElement=\\"SequenceFlow_1w1cqrc\\">
        <di:waypoint x=\\"533\\" y=\\"241\\" />
        <di:waypoint x=\\"583\\" y=\\"241\\" />
      </bpmndi:BPMNEdge>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn:definitions>
",
  "resourceName": "MESSAGE_EVENT.bpmn",
  "version": 1,
  "workflowKey": "2251799813685250",
}
`;

exports[`Zeebe Worker should get workflows 1`] = `
Object {
  "items": Array [
    Object {
      "bpmnProcessId": "MESSAGE_EVENT",
      "resourceName": "MESSAGE_EVENT.bpmn",
      "version": 1,
      "workflowKey": "2251799813685250",
    },
  ],
  "paging": Object {
    "from": 0,
    "size": 500,
    "totalCount": 1,
  },
}
`;

exports[`Zeebe Worker should limit the number of document 1`] = `
Object {
  "items": Array [
    Object {
      "bpmnProcessId": "test-process",
      "resourceName": "test.bpmn",
      "version": 2,
      "workflowKey": "2251799813709933",
    },
    Object {
      "bpmnProcessId": "test-process",
      "resourceName": "test.bpmn",
      "version": 3,
      "workflowKey": "2251799813719839",
    },
    Object {
      "bpmnProcessId": "test-process",
      "resourceName": "test.bpmn",
      "version": 11,
      "workflowKey": "2251799813720710",
    },
    Object {
      "bpmnProcessId": "test-process",
      "resourceName": "test.bpmn",
      "version": 10,
      "workflowKey": "2251799813720691",
    },
    Object {
      "bpmnProcessId": "test-process",
      "resourceName": "test.bpmn",
      "version": 14,
      "workflowKey": "2251799813720733",
    },
  ],
  "paging": Object {
    "from": 0,
    "size": 5,
    "totalCount": 15,
  },
}
`;

exports[`Zeebe Worker should limit the number of document and search by bpmnProcessId 1`] = `
Object {
  "items": Array [
    Object {
      "bpmnProcessId": "MESSAGE_EVENT",
      "resourceName": "MESSAGE_EVENT.bpmn",
      "version": 1,
      "workflowKey": "2251799813685250",
    },
  ],
  "paging": Object {
    "from": 0,
    "size": 5,
    "totalCount": 1,
  },
}
`;

exports[`Zeebe Worker should limit the number of document and skip one document 1`] = `
Object {
  "items": Array [
    Object {
      "bpmnProcessId": "test-process",
      "resourceName": "test.bpmn",
      "version": 3,
      "workflowKey": "2251799813719839",
    },
    Object {
      "bpmnProcessId": "test-process",
      "resourceName": "test.bpmn",
      "version": 11,
      "workflowKey": "2251799813720710",
    },
    Object {
      "bpmnProcessId": "test-process",
      "resourceName": "test.bpmn",
      "version": 10,
      "workflowKey": "2251799813720691",
    },
    Object {
      "bpmnProcessId": "test-process",
      "resourceName": "test.bpmn",
      "version": 14,
      "workflowKey": "2251799813720733",
    },
  ],
  "paging": Object {
    "from": 1,
    "size": 4,
    "totalCount": 15,
  },
}
`;
